proposals/nnnn-move-function.md

```diff
@@ -0,0 +1,455 @@
+
+This follows from move being applied to the binding (`x`), not the value in the
+binding (the value returned from `getValue()`).
+
+We also support applying the move operation to consuming function arguments:
+
+```
+func f(_ x: __owned SomeClassType) {
+    let _ = move(x)
+    useX(x) // !! Error! Use of x after move
+}
+```
+
+Normal arguments are passed by borrow, meaning that the lifetime of the value
+is managed by the caller. Although we could allow `move` on these arguments,
+shortening the syntactic lifetime of the variable, doing so would have no
+practical effect on the value's lifetime at runtime, so we choose to leave this
+disallowed for now.
+
+On the other hand, one can `move` out of an `inout`
+function argument. Like a `var`, the `inout` argument can be reassigned after
+being moved from and used again; however, since the final value of an
+`inout` argument is passed back to the caller, an `inout` argument *must* be
+reassigned by the callee before it returns:
+
+```
+func f(_ buffer: inout Buffer) { // error: 'buffer' not reinitialized after move!
+  let b = move(buffer)           // note: move was here
+  b.deinitialize()
+  ... write code ...
+}                                // note: return without reassigning inout argument `buffer`
+```
+
+But we can to re-initialize `buffer` by writing the following code:
+
+```
+func f(_ buffer: inout Buffer) {
+  let b = move(buffer)
+  b.deinitialize()
+  // ... write code ...
+  // We re-initialized buffer before end of function so the checker is satisfied
+  buffer = getNewInstance()
+}
+```
+
+Move analysis understands that `defer` statements run before every scope exit,
+so `defer` can also be used to reinitialize an `inout` or `var` after a move.
+So we can also write the above as:
+
+```
+func f(_ buffer: inout Buffer) {
+  let b = move(buffer)
+  // Ensure the buffer is reinitialized before we exit.
+  defer { buffer = getNewInstance() }
+  b.deinitialize()
+  // ... write code ...
+}
+```
+
+
+In the future, we may add support for globals and stored properties, although
+doing so in full generality would require dynamic enforcement in addition to
+static checking to ensure that shared state is not read from once it is moved,
+similar to how we need to dynamically enforce exclusivity when accessing
+globals and class stored properties. For now, `move` will raise an error when
+applied to something we cannot statically analyze, so this code:
+
+```
+var global = SomeClassType()
+func f() {
+  let _ = move(global)
+}
+```
+
+would complain:
+
+```
+test.swift:9:11: error: move applied to value that the compiler does not support checking
+  let _ = move(global)
+          ^
+```
+
+## Detailed design
+
+We define move as follows:
+
+```
+/// This function ends the lifetime of the passed in binding.
+@_transparent
+@alwaysEmitIntoClient
+func move<T>(_ t: __owned T) -> T {
+  Builtin.move(t)
+}
+```
+
+Builtin.move is a hook in the compiler to force emission of special SIL "move"
```

Comment:
I think you should delete the rest of this section from this point.

At most, I would like to see a brief paragraph that says in essence that "Builtin.move() causes the compiler to perform additional data flow analysis to verify that the argument is not in fact used after this point."  The details of how that is implemented in this particular compiler is not relevant; a different compiler may implement it in very different ways.